### PR TITLE
feat(analytics): FRM Analytics

### PR DESCRIPTION
**Pull request by tsdk02**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Added frm dashboard metrics and filters.
Filters based on:
- frm_status
- frm_name
- frm_transaction_type

Metrics calculated:
- frm_triggered_attempts
- frm_blocked_rate
### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Can get analytics insights based on frm

## How did you test it?
Below is the collection for testing :
[frms.postman_collection.json](https://github.com/user-attachments/files/15897457/frms.postman_collection.json)

Tested using local data by sending cURL requests through Postman.
Screenshot 2024-06-05 at 11 57 32 AM
Screenshot 2024-06-05 at 11 57 49 AM
Screenshot 2024-06-05 at 11 58 01 AM
Screenshot 2024-06-05 at 11 58 35 AM
## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Comment by ivor-juspay** [comment]
image
all dimensions are null. Is the deserialization to the internal struct happening correctly?

File: crates/analytics/src/frm/core.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    req: GetFrmFilterRequest,
+    merchant_id: &String,
+) -> AnalyticsResult<FrmFiltersResponse> {
+    let mut res = FrmFiltersResponse::default();
+    for dim in req.group_by_names {
+        let values = match pool {
+            AnalyticsProvider::Clickhouse(_pool) => Err(FiltersError::NotImplemented(
+                "Frm not implemented for Clickhouse",
+            ))
+            .attach_printable("Clickhouse Analytics is not implemented for Clickhouse"),
```
Comment (ivor-juspay) [inline crates/analytics/src/frm/core.rs:126]: Clickhouse Analytics is not implemented for FRM

**Comment by tsdk02** [comment]
Adding a curl req/ resp with some dimensions in groupByNames clause. It works as expected.
Link to comment: https://github.com/juspay/hyperswitch/pull/4880#issuecomment-2149343307
Screenshot 2024-06-06 at 5 05 41 PM

**Review by lsampras — APPROVED** [review]

File: crates/analytics/src/frm/core.rs [hunk 2]
```diff
@@ -120,16 +118,61 @@ pub async fn get_filters(
...
+    let sqlx_result = get_frm_filter_for_dimension(
+        dim,
+        merchant_id,
+        &req.time_range,
+        sqlx_pool,
+    )
+    .await;
+    match (&sqlx_result, &ckh_result) {
+        (Ok(ref sqlx_res), Ok(ref ckh_res)) if sqlx_res != ckh_res => {
+            router_env::logger::error!(clickhouse_result=?ckh_res, postgres_result=?sqlx_res, "Mismatch between clickhouse & postgres refunds analytics filters")
```
Comment (ivor-juspay) [inline crates/analytics/src/frm/core.rs:146]: Mismatch between clickhouse & postgres frm analytics

File: crates/router/src/db.rs [hunk 3]
```diff
@@ -255,36 +261,75 @@ impl RequestIdStore for KafkaStore {
...
-            .await
+            .await?;
+
+        if let Some(fraud_check) = frm.clone() {
+            if let Err(er) = self
+                .kafka_producer
+                .log_fraud_check(&fraud_check, None, self.tenant_id.clone())
+                .await
+            {
+                logger::error!(message="Failed to log analytics event for payment attempt {attempt:?}", error_message=?er);
```
Comment (ivor-juspay) [inline crates/router/src/db.rs:328]: frm

File: crates/router/src/services/kafka/fraud_check.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            payment_capture_method: check.payment_capture_method,
+        }
+    }
+}
+
+impl<'a> super::KafkaMessage for KafkaFraudCheck<'a> {
+    fn key(&self) -> String {
+        format!(
+            "{}_{}_{}",
+            self.merchant_id, self.payment_id, self.attempt_id
```
Comment (ivor-juspay) [inline crates/router/src/services/kafka/fraud_check.rs:60]: is frm_id required in the key?

File: crates/router/src/services/kafka/fraud_check_event.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
+// use diesel_models::enums as storage_enums;
+use diesel_models::{
+    enums as storage_enums,
+    enums::{FraudCheckLastStep, FraudCheckStatus, FraudCheckType},
+    fraud_check::FraudCheck,
+};
+use time::OffsetDateTime;
+
+#[derive(serde::Serialize, Debug)]
+pub struct KafkaFraudCheckEvent<'a> {
```
Comment (ivor-juspay) [inline crates/router/src/services/kafka/fraud_check_event.rs:10]: can use KafkaFraudCheckEvent everywhere and remove KafkaFraudCheck

Comment (Abhitator216) [reply crates/router/src/services/kafka/fraud_check_event.rs:10]: Actually i followed the same structure of event generation in kafka

File: crates/router/src/services/kafka.rs [hunk 6]
```diff
@@ -306,6 +313,34 @@ impl KafkaProducer {
...
+            .attach_printable_lazy(|| {
+                format!("Failed to add negative attempt event {negative_event:?}")
+            })?;
+        };
+
+        self.log_event(&KafkaEvent::new(
+            &KafkaFraudCheck::from_storage(attempt),
+            tenant_id.clone(),
+        ))
+        .attach_printable_lazy(|| format!("Failed to add positive attempt event {attempt:?}"))?;
```
Comment (lsampras) [inline crates/router/src/services/kafka.rs:336]:
```suggestion
        .attach_printable_lazy(|| format!("Failed to add positive fraud check event {attempt:?}"))?;
```

File: crates/router/src/services/kafka.rs [hunk 7]
```diff
@@ -306,6 +313,34 @@ impl KafkaProducer {
...
+        old_attempt: Option<FraudCheck>,
+        tenant_id: TenantID,
+    ) -> MQResult<()> {
+        if let Some(negative_event) = old_attempt {
+            self.log_event(&KafkaEvent::old(
+                &KafkaFraudCheck::from_storage(&negative_event),
+                tenant_id.clone(),
+            ))
+            .attach_printable_lazy(|| {
+                format!("Failed to add negative attempt event {negative_event:?}")
```
Comment (lsampras) [inline crates/router/src/services/kafka.rs:328]:
```suggestion
                format!("Failed to add negative fraud check event {negative_event:?}")
```

File: crates/router/src/services/kafka.rs [hunk 8]
```diff
@@ -306,6 +313,34 @@ impl KafkaProducer {
...
+            &KafkaFraudCheck::from_storage(attempt),
+            tenant_id.clone(),
+        ))
+        .attach_printable_lazy(|| format!("Failed to add positive attempt event {attempt:?}"))?;
+
+        self.log_event(&KafkaConsolidatedEvent::new(
+            &KafkaFraudCheckEvent::from_storage(attempt),
+            tenant_id.clone(),
+        ))
+        .attach_printable_lazy(|| format!("Failed to add consolidated attempt event {attempt:?}"))
```
Comment (lsampras) [inline crates/router/src/services/kafka.rs:342]:
```suggestion
        .attach_printable_lazy(|| format!("Failed to add consolidated fraud check event {attempt:?}"))
```

File: crates/router/src/services/kafka/fraud_check_event.rs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            payment_capture_method: check.payment_capture_method,
+        }
+    }
+}
+
+impl<'a> super::KafkaMessage for KafkaFraudCheckEvent<'a> {
+    fn key(&self) -> String {
+        format!(
+            "{}_{}_{}",
+            self.merchant_id, self.payment_id, self.attempt_id
```
Comment (lsampras) [inline crates/router/src/services/kafka/fraud_check_event.rs:60]: this would also include frm_id

File: crates/router/src/services/kafka/fraud_check_event.rs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    pub frm_name: &'a String,
+    pub frm_transaction_id: Option<&'a String>,
+    pub frm_transaction_type: FraudCheckType,
+    pub frm_status: FraudCheckStatus,
+    pub frm_score: Option<i32>,
+    pub frm_reason: Option<serde_json::Value>,
+    pub frm_error: Option<&'a String>,
+    pub payment_details: Option<serde_json::Value>,
+    pub metadata: Option<serde_json::Value>,
+    #[serde(with = "time::serde::timestamp")]
```
Comment (lsampras) [inline crates/router/src/services/kafka/fraud_check_event.rs:26]: the events format uses milliseconds refer to the `dispute_event.rs`.

File: crates/router/src/db/kafka_store.rs [hunk 11]
```diff
@@ -2036,6 +2040,83 @@ impl RefundInterface for KafkaStore {
...
+//                 .await
+//             {
+//                 logger::error!(message="Failed to log analytics event for payment attempt {attempt:?}", error_message=?er);
+//             }
+//         }
+//         Ok(frm)
+//         // Ok(frm.clone()) // one error for partial borrowing - borrow of partially moved value: `frm`
+//     }
+// }
+
```
Comment (lsampras) [inline crates/router/src/db/kafka_store.rs:2119]: 😞

File: crates/router/src/db/fraud_check.rs [hunk 12]
```diff
@@ -117,32 +117,32 @@ impl FraudCheckInterface for MockDb {
...
-}
+//     #[instrument(skip_all)]
+//     async fn find_fraud_check_by_payment_id(
+//         &self,
+//         _payment_id: String,
+//         _merchant_id: String,
+//     ) -> CustomResult<FraudCheck, errors::StorageError> {
+//         Err(errors::StorageError::MockDbError)?
+//     }
+// }
```
Comment (lsampras) [inline crates/router/src/db/fraud_check.rs:148]: 😞

File: crates/router/src/db.rs [hunk 13]
```diff
@@ -255,36 +261,75 @@ impl RequestIdStore for KafkaStore {
...
+        new: storage::FraudCheckNew,
     ) -> CustomResult<FraudCheck, StorageError> {
-        self.diesel_store.insert_fraud_check_response(new).await
+        let frm = self.diesel_store.insert_fraud_check_response(new).await?;
+        if let Err(er) = self
+            .kafka_producer
+            .log_fraud_check(&frm, None, self.tenant_id.clone())
+            .await
+        {
+            logger::error!(message = "Failed to log analytics event for payment attempt", error_message = ?er);
```
Comment (lsampras) [inline crates/router/src/db.rs:272]:
```suggestion
            logger::error!(message = "Failed to log analytics event for fraud check", error_message = ?er);
```

File: crates/router/src/db.rs [hunk 14]
```diff
@@ -255,36 +261,75 @@ impl RequestIdStore for KafkaStore {
...
+        let frm = self
+            .diesel_store
+            .update_fraud_check_response_with_attempt_id(this, fraud_check)
+            .await?;
+        if let Err(er) = self
+            .kafka_producer
+            .log_fraud_check(&frm, None, self.tenant_id.clone())
             .await
+        {
+            logger::error!(message="Failed to log analytics event for payment attempt {attempt:?}", error_message=?er)
```
Comment (lsampras) [inline crates/router/src/db.rs:290]:
```suggestion
            logger::error!(message="Failed to log analytics event for fraud check {frm:?}", error_message=?er)
```

File: crates/api_models/Cargo.toml [hunk 15]
```diff
@@ -35,7 +35,7 @@ url = { version = "2.5.0", features = ["serde"] }
 utoipa = { version = "4.2.0", features = ["preserve_order", "preserve_path_order"] }
 frunk = "0.4.2"
 frunk_core = "0.4.2"
-
+diesel_models = { version = "0.1.0", path = "../diesel_models" }
```
Comment (lsampras) [inline crates/api_models/Cargo.toml:38]: Ideally this import shouldn't be there.
we could instead move these enums to common_enums maybe

File: crates/analytics/src/frm/accumulator.rs [hunk 16]
```diff
@@ -0,0 +1,78 @@
+use api_models::analytics::frm::FrmMetricsBucketValue;
+use diesel_models::enums as storage_enums;
```
Comment (lsampras) [inline crates/analytics/src/frm/accumulator.rs:2]: 🚨 
importing diesel models would cause higher compilation times

**Review by lsampras — CHANGES_REQUESTED** [review]

File: crates/router/src/db.rs (hunk 14 above)
Comment (lsampras) [reply crates/router/src/db.rs:290]: tbh Idk how does this even compile? since it relies on a variable called attempt which isn't present

**Comment by lsampras** [comment]
Also need to change `config/deployments/env_specific.toml`

File: crates/analytics/src/frm/core.rs (hunk 2 above)
Comment (Abhitator216) [reply crates/analytics/src/frm/core.rs]: Changes Done

File: crates/router/src/services/kafka/fraud_check.rs (hunk 4 above)
Comment (Abhitator216) [reply crates/router/src/services/kafka/fraud_check.rs]: Done

File: crates/router/src/db.rs (hunk 3 above)
Comment (Abhitator216) [reply crates/router/src/db.rs]: Done

File: config/development.toml [hunk 17]
```diff
@@ -584,6 +584,7 @@ source = "logs"
 
 [events.kafka]
 brokers = ["localhost:9092"]
+fraud_check_analytics_topic= "hyperswitch-fraud-check-events"
```
Comment (ivor-juspay) [inline config/development.toml:587]: raise hyperswitch-cloud PR for adding these topics to envs
ex: https://github.com/juspay/hyperswitch-cloud/pull/5266/

Comment (Abhitator216) [reply config/development.toml:589]: Done

File: crates/analytics/src/frm/accumulator.rs (hunk 16 above)
Comment (Abhitator216) [reply crates/analytics/src/frm/accumulator.rs:2]: Done

File: crates/analytics/src/frm/core.rs [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+pub async fn get_filters(
+    pool: &AnalyticsProvider,
+    req: GetFrmFilterRequest,
+    merchant_id: &String,
+) -> AnalyticsResult<FrmFiltersResponse> {
+    let mut res = FrmFiltersResponse::default();
+    for dim in req.group_by_names {
+        let values = match pool {
```
[the remark this answers is not in the document]
Comment (Abhitator216) [reply crates/analytics/src/frm/core.rs:120]: Okay

File: crates/api_models/Cargo.toml (hunk 15 above)
Comment (Abhitator216) [reply crates/api_models/Cargo.toml:38]: Done

File: crates/router/src/db.rs (hunk 13 above)
Comment (Abhitator216) [reply crates/router/src/db.rs]: Done

File: crates/router/src/db.rs (hunk 14 above)
Comment (Abhitator216) [reply crates/router/src/db.rs]: Done

File: crates/common_enums/src/enums.rs [hunk 19]
```diff
@@ -231,6 +232,30 @@ pub enum AuthenticationType {
...
+    Copy,
+    Debug,
+    Default,
+    Eq,
+    PartialEq,
+    serde::Serialize,
+    serde::Deserialize,
+    strum::Display,
+    strum::EnumString,
+    frunk::LabelledGeneric,
```
Comment (ivor-juspay) [inline crates/common_enums/src/enums.rs:246]: is this derive required?

File: crates/router/src/db/fraud_check.rs [hunk 20]
```diff
@@ -116,33 +116,3 @@ impl FraudCheckInterface for MockDb {
         Err(errors::StorageError::MockDbError)?
     }
 }
-
-#[cfg(feature = "kafka_events")]
-#[async_trait::async_trait]
-impl FraudCheckInterface for super::KafkaStore {
```
Comment (ivor-juspay) [inline crates/router/src/db/fraud_check.rs:122]: why is this removed?

Comment (Abhitator216) [reply crates/router/src/db/fraud_check.rs:122]: `#[cfg(feature = "kafka_events")]`
It used to be a feature flag, but now it is not being used.

File: crates/common_enums/src/enums.rs (hunk 19 above)
Comment (Abhitator216) [reply crates/common_enums/src/enums.rs:246]: > is this derive required?
Because of this comment by @lsampras  we changed the location of enums to `common_enums`
https://github.com/juspay/hyperswitch/pull/4880#discussion_r1637501479

File: crates/router/src/db/fraud_check.rs (hunk 20 above)
Comment (Abhitator216) [reply crates/router/src/db/fraud_check.rs:122]: > why is this removed?

It used to be a feature flag, but now it is not being used.

**Review by lsampras — APPROVED** [review]

**Review by ShivanshMathurJuspay — DISMISSED** [review]

**Review by Abhitator216 — DISMISSED** [review]

**Review by ivor-juspay — APPROVED** [review]

**Review by lsampras — DISMISSED** [review]

**Review by Abhitator216 — DISMISSED** [review]

**Review by Abhitator216 — APPROVED** [review]

**Review by Abhitator216 — APPROVED** [review]

**Review by ShivanshMathurJuspay — APPROVED** [review]

**Review by ivor-juspay — APPROVED** [review]

**Review by likhinbopanna — APPROVED** [review]

**Review by lsampras — APPROVED** [review]
